### PR TITLE
Feature: Nested aggregations on interface

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR introduces nested aggregations on interfaces when the `experimental` flag is passed as `true` to the Neo4jGraphQL constructor.

Given the following type definitions, we can perform aggregations on actedIn where the relationship target is an interface (`Production`). The fields available for the aggregation operation are those in the interface (`title` and `cost` in the example).

```gql
interface Production {
    title: String!
    cost: Float!
}

type Movie implements Production {
    title: String!
    cost: Float!
    runtime: Int!
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN, properties: "ActedIn")
}

type Series implements Production {
    title: String!
    cost: Float!
    episodes: Int!
}

interface ActedIn @relationshipProperties {
    screenTime: Int!
}

type Actor {
    name: String!
    actedIn: [Production!]! @relationship(type: "ACTED_IN", direction: OUT, properties: "ActedIn")
}
```

An example aggregation that is now possible:
```gql
actors {
    actedInAggregate {
        count
        node {
            cost {
                max
            }
        }
    }
}
```

## Complexity

Medium

There have been relatively large changes made to the `OperationFactory` which will be further worked on when we continue interface aggregation work.

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4257.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 40 | 59 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 85 | 96 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 58 | 77 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3557 | 3750 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3189 | 3140 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 51 | 62 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 67 | 58 | 2174 |
|  aggregations.DeeplyNestedCount | 11062443 | 11062443 | 5363 | 5429 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 151 | 196 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 149 | 75 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 249 | 257 | 3003 |
|  connections.Connection | 14082 | 14082 | 74 | 74 | 2174 |
|  connections.NestedConnection | 41118 | 41118 | 105 | 150 | 4516 |
|  create.SimpleMutation | 7 | 7 | 63 | 76 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 58 | 32 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 311 | 237 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 223 | 295 | 1179 |
|  2871.NestedRelationshipFilter | 23066 | 23081 | 64 | 86 | 4414 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 48 | 64 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 99 | 144 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 45 | 56 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 63 | 85 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 28 | 26 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 46 | 60 | 2174 |
|  query.QueryWhere | 9711 | 9718 | 42 | 53 | 2166 |
|  query.SimpleQueryWithNestedWhere | 9889 | 9896 | 60 | 66 | 2166 |
|  query.Nested | 10096041 | 10096041 | 10182 | 10365 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 10307 | 10311 | 2004000 |
|  query.OrFilterOnRelationships | 42482 | 42184 | 181 | 241 | 1791 |
|  query.OrFilterOnRelationshipsAndNested | 36831 | 35069 | 235 | 339 | 1791 |
|  query.QueryWithNestedIn | 13558 | 13464 | 67 | 83 | 1268 |
|  query.NestedConnectionWhere | 9834 | 9834 | 95 | 85 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9851 | 9899 | 86 | 105 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18996 | 19110 | 189 | 184 | 1552 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 198 | 286 | 1134 |
|  query.Fulltext | 80 | 80 | 37 | 40 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 53 | 57 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14092 | 14092 | 61 | 61 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14092 | 14092 | 78 | 94 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14238 | 14219 | 77 | 88 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14238 | 14219 | 96 | 112 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14591 | 14742 | 115 | 137 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14378 | 14566 | 118 | 280 | 2174 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 74 | 156 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14464 | 14530 | 100 | 192 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 111 | 68 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 114 | 113 | 4516 |
|  sorting.ConnectionWithSort | 3284 | 3284 | 82 | 106 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 112 | 74 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 122 | 79 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 331 | 363 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 291 | 354 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 116 | 135 | 2002 |

</details>
        
Old Schema Generation: 51.347s
Schema Generation: 47.659s
Old Subgraph Schema Generation: 1:05.434 (m:ss.mmm)
Subgraph Schema Generation: 1:06.833 (m:ss.mmm)

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                f.getAggregationProjection(nestedContext.returnVariable, nestedContext.returnVariable)
+            );
+        });
+
+        if (nodeMap.size > 0) {
+            aggregationProjectionMap.set("node", nodeMap);
+        }
+        // if (edgeMap.size > 0) {
+        //     aggregationProjectionMap.set("edge", edgeMap);
+        // }
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts:145]: :eyes:

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            this.pagination,
+            ...this.children,
+        ]);
+    }
+
+    protected getSortFields(context: QueryASTContext, target: Cypher.Variable): SortField[] {
+        return this.sortFields.flatMap((sf) => sf.getSortFields(context, target, false));
+    }
+
+    public transpile({ context }: OperationTranspileOptions): OperationTranspileResult {
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts:84]: We may want to consider moving some parts of this into separate private methods. For instance the fieldSubqueries calculation

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationPartial.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import { createNodeFromEntity, createRelationshipFromEntity } from "../../../utils/create-node-from-entity";
+import { wrapSubqueriesInCypherCalls } from "../../../utils/wrap-subquery-in-calls";
+import type { QueryASTContext } from "../../QueryASTContext";
+import { QueryASTNode } from "../../QueryASTNode";
+
+export class CompositeAggregationPartial extends QueryASTNode {
+    public readonly entity: RelationshipAdapter;
+    public readonly target: ConcreteEntityAdapter;
+    protected directed: boolean;
+    // private returnVariable: Cypher.Variable;
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationPartial.ts:32]:
```suggestion
```

**Review by angrykoala — CHANGES_REQUESTED** [review]
Looks good overall, some commented code left that needs to be removed before merging

File: packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts (hunk 1 above)
Comment (mjfwebb) [reply packages/graphql/src/translate/queryAST/ast/operations/composite/CompositeAggregationOperation.ts:144]: Will address edges in followup

**Review by angrykoala — APPROVED** [review]

**Pull request merged**